### PR TITLE
*: use column id to get stats instead of column offset

### PR DESCRIPTION
**Pull request by hanfei1991**
Now we assume that the order of columns in stats matches the column offsets. It will return wrong stats sometimes.
If we analyze when a table is (c1, c2) and then drop c1, the c2's offset will become 0. But in stats the position of 0 is c1. I have added this case to test.
@shenli @coocood @zimulala @lamxTyler PTAL

### CODE REVIEW

**Comment by lamxTyler** [comment]
I tried the test in the old version, it still passes. Could you provide another test?

**Comment by hanfei1991** [comment]
@lamxTyler Indeed, I cann't reproduce this bug... We can avoid this bug  by using pseudo table ... but this will be a prob when we use pseudo column.

**Comment by lamxTyler** [comment]
I don't understand why this bug will occur. Currently, if the only change ever happened is drop columns, then another columns will automatically fill the hole, because we just append the column to the stats. In the example that you provided, stats in postion 0 will be c2, not c1.

**Comment by hanfei1991** [comment]
There are two cases that may trigger this bug in the future:

1. write stats with old schema and load with new schema. In future we may support swith the offsets of two columns, it's a nightmare
2. get stats with new schema. Now If we observe ddl changed, we will return a pseudo table, it's unreasonable. We'd better return the stats and hold the exception during range estimating. At that time we must process the column ID with wrong offset

btw, column ID will never be reused for the same table, so it is safe.

**Comment by lamxTyler** [comment]
As long as the columns id in the table info is a subsequence of the `histId` that we get from store, there won't be a problem. That's all I know from the code.

**Comment by lamxTyler** [comment]
The two case that you provide will cause this bug. However, Why don't we set the stats according to the schema, not the order in the stats we get from the store? This won't be any problem, and we need only change one function `TableStatsFromStorage`.

**Comment by hanfei1991** [comment]
Adjusting the order of columns when load stas from kv can avoid case 1. But case 2 cannot be avoided, we are not sure if the comming tableInfo ID has the same offset with the stats instead of checking columns in slice one by one.

**Comment by hanfei1991** [comment]
The case 2 happens when load stats and ddl happened. In future we will update the stats when we have done the ddl, but we cannot adjust the order of records in TiKV. And there still exists a gap when the ddl updates but stats doesn't update.

**Comment by lamxTyler** [comment]
Put it in the pseudo code.
```
foreach col in tblInfo.Columns:
     if col stats does not exist in TiKV
         append psudo columns;
    else 
         append col stats.
```
Why there would be a problem?

**Comment by hanfei1991** [comment]
@lamxTyler This method can only resolve the case 1 instead of case 2. Case 2 will get the stats from map (not load stats from kv) with new schema whose offsets are wrong.

**Comment by lamxTyler** [comment]
Got it.

**Comment by lamxTyler** [comment]
Then the test can be updated using the case 2?

**Comment by hanfei1991** [comment]
@lamxTyler I'm sorry we can't, because if the tableInfo doesn't match the old one, it will return a pseudo table...

File: plan/planbuilder.go [hunk 1]
```diff
@@ -361,34 +361,33 @@ func (b *planBuilder) buildAdmin(as *ast.AdminStmt) Plan {
 	return p
 }
 
-// getColumnOffsets returns the offsets of index columns, normal columns and primary key with integer type.
-func getColumnOffsets(tn *ast.TableName) (indexOffsets []int, columnOffsets []int, pkOffset int) {
+// getColumnIDs returns the offsets of index columns, normal columns and primary key with integer type.
```
Comment (lamxTyler) [inline plan/planbuilder.go]: The comments should be updated.

File: statistics/builder.go [hunk 2]
```diff
@@ -35,21 +35,21 @@ type Builder struct {
...
 	ColumnSamples [][]types.Datum  // ColumnSamples is the sample of columns.
-	ColOffsets    []int            // ColOffsets is the offset of columns in the table.
+	ColIDs        []int64          // ColOffsets is the offset of columns in the table.
 	ColNDVs       []int64          // ColNDVs is the NDV of columns.
 	IdxRecords    []ast.RecordSet  // IdxRecords is the record set of index columns.
-	IdxOffsets    []int            // IdxOffsets is the offset of indices in the table.
+	IdxIDs        []int64          // IdxOffsets is the offset of indices in the table.
 	PkRecords     ast.RecordSet    // PkRecords is the record set of primary key of integer type.
-	PkOffset      int              // PkOffset is the offset of primary key of integer type in the table.
+	PkID          int64            // PkOffset is the offset of primary key of integer type in the table.
```
Comment (lamxTyler) [inline statistics/builder.go]: Updated the comment.

File: statistics/builder.go [hunk 3]
```diff
@@ -114,25 +114,25 @@ func (b *Builder) NewTable() (*Table, error) {
...
 		return t, nil
 	}
-	err := b.splitAndConcurrentBuild(t, b.ColOffsets, false)
+	err := b.splitAndConcurrentBuild(t, b.ColIDs, false)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	if b.PkOffset != -1 {
-		err := t.build4SortedColumn(b.Ctx.GetSessionVars().StmtCtx, b.PkOffset, b.PkRecords, b.NumBuckets, true)
+	if b.PkID != -1 {
```
Comment (coocood) [inline statistics/builder.go]: We don't need to set and check PKID to `-1` now.

Comment (hanfei1991) [inline statistics/builder.go]: Why not ?

File: statistics/statscache_test.go [hunk 4]
```diff
@@ -194,6 +194,40 @@ func (s *testStatsCacheSuite) TestEmptyTable(c *C) {
 	c.Assert(count, Equals, int64(3333333))
 }
 
+func (s *testStatsCacheSuite) TestColumnOffset(c *C) {
```
Comment (coocood) [inline statistics/statscache_test.go]: `TestColumnIDs`?

**Comment by hanfei1991** [comment]
@coocood @lamxTyler PTAL

File: statistics/builder.go (hunk 3 above)
Comment (coocood) [inline statistics/builder.go]: Then we don't need to worry about lost the initialization during a refactor.

**Comment by lamxTyler** [comment]
LGTM

File: statistics/table.go [hunk 5]
```diff
@@ -42,8 +43,9 @@ const (
 // Table represents statistics for a table.
 type Table struct {
 	Info    *model.TableInfo
-	Columns []*Column
-	Indices []*Column
+	m       sync.Mutex
```
Comment (coocood) [inline statistics/table.go]: We should make the Table read-only to avoid lock cost.

Comment (lamxTyler) [inline statistics/table.go]: This lock is only used when build columns, and after that it is not used anymore.

File: plan/statistics_helper.go [hunk 6]
```diff
@@ -116,8 +116,15 @@ func getPseudoRowCountByIndexRanges(sc *variable.StatementContext, statsTbl *sta
 		}
 		l := indexRange.LowVal[i]
 		r := indexRange.HighVal[i]
-		offset := indexInfo.Columns[i].Offset
-		rowCount, err := getRowCountByRange(sc, statsTbl.Count, statsTbl.Columns[offset], l, r)
+		name := indexInfo.Columns[i].Name
```
Comment (coocood) [inline plan/statistics_helper.go]: We can safely get the index column by 
```
colOffset := indexInfo.Columns[i].Offset
table.Columns[collOffset]
```

**Comment by hanfei1991** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**